### ISSUE TITLE
Automated submission using GitHub actions

### ISSUE THREAD

**Issue by seabbs** [epiforecasts/covid19-forecast-hub-europe#15]
If wanting to offer automated submission via GitHub actions (which I think it would be nice to support as it reduces overhead for groups). We have a R based setup here that could be generalised.

Action: https://github.com/KITmetricslab/covid19-forecast-hub-de/blob/master/.github/workflows/epiforecasts.yml
R extraction script (specific to Epiforecasts but should be easily generalisable): https://github.com/KITmetricslab/covid19-forecast-hub-de/blob/master/code/auto_download/auto_download_epiforecasts.R

Related to #14 .

**Comment by kathsherratt** [comment]
Sounds good, thanks for suggesting this and linking! 

Do you think this would significantly reduce the amount of knowledge/comfort with Github needed to submit? If so we should prioritise putting this in place to make this all as accessible as possible.

However if it's a handy add on for teams which will already be comfortable with forks/PRs then perhaps one for the to-do list as or after we have got up and running.

**Comment by seabbs** [comment]
If the tools are generalised then it would just require them to have a public folder somewhere with the forecasts in the correct format. 

So you'd just pass a list as here: https://github.com/KITmetricslab/covid19-forecast-hub-de/blob/ac8410fab73f03a9ccd60a0e345e3135dc263a09/code/auto_download/auto_download_epiforecasts.R#L111

+ a target URL. The tricky bit is if they produce forecasts on varying dates but that issue can be on them. 

The actual GitHub action could then be modified from a template. 

Personally, I feel quite fatigued by the PR set up and I can imagine stopping submitting if there weren't other people available/an automated approach. Especially when you consider the benefit teams typically get from submitting in the first place. 

TLDR: I don't think it is critical but it would be nice to have. Manually submitting gets very dull and seems like a pretty pointless barrier to entry.

**Comment by kathsherratt** [comment]
I'd probably agree - the PR system seems like a whole extra sequence of github hoops to jump through - even if pretty easy it takes explaining and I think might be one of those minor things that makes it much harder to share work around a team. In general keen to reduce barriers as much as possible.

I am pretty new to working with GH actions but learning. When you have had a look at the validation issue with the epiforecasts german hub submission then I can try and replicate here.

@sbfnk would you be happy with this - noting it was not on the priority list in last thursday's discussion?

**Comment by seabbs** [comment]
In https://github.com/KITmetricslab/covid19-forecast-hub-de/issues/577 there are some valid issues with automation so be worth seeing what comes of that. It probably comes down to where I/we come down on the spectrum of fatigue for people submitting to hubs.

**Comment by seabbs** [comment]
This https://github.com/KITmetricslab/covid19-forecast-hub-de/issues/577#issuecomment-763098991 seems to be the broad solution to making PR submission automated. I'd suggest having a brief wiki entry etc outlining how to automate the expected submission format. 

If doesn't sound like the German hub team is interested in building out a solution for those submitting to their hub (which is fair enough). I'll work out something that works for us in the German and US hub and write some docs on how others could adapt it.

**Comment by seabbs** [comment]
Looks like `hub` has been superseded by GitHub command line tools (https://cli.github.com/manual/gh_pr_create).

**Comment by jbracher** [comment]
I'll just add my two cents on this based on my experience from the German and US projects.

- **Keep the work away from your team.** If people want to automate things, fine. But let them do it in their own repos or wherever and interact with your repo in the most standardized way possible. Anything that takes place within your repo or is provided as an additional infrastructure by you will cause you work. You'll hopefully have a lot of teams, and you're a small group. You don't want to help them set up stuff inside your repo.
- People get the format wrong all the time. We've maybe had two teams who managed to submit a valid csv file the first time they tried. Even among the teams who have done this for a while we catch one or two faulty submissions each week. So there is a point in these validation checks and letting anybody other than your core team push directly to the main branch will break things sooner or later.
- If I had to set up things again I don't think I would keep the codes which get in forecasts automatically form other sources in the main repo (even though they only update data via PR). This processing is a task which I consider distinct from the main purpose of the Hub (which imo is hosting, showing, evaluating and combining forecasts).
- I would advise against having anybody other than you run codes inside your repo. We allowed Sam to do that because we know him, but it's impossible to check what these scripts are actually doing. You'll want to keep things simple and controllable.
- Don't overestimate how good people are at automating things. As far as we know no other team has attempted to automate their submissions.
- A more general question is whether forecasts that are (or can be) fully automated are what ECDC will be most interested in. There is no hard evidence that models involving human input or tuning really do better, and to be fair, EpiNow2 has shown good relative performance in the German/Polish project. But I think there are natural limits to such approaches. I don't expect them to see changes in trends coming (e.g. due to effects of known interventions kicking is). But that's exactly the kind of things "consumers" of forecasts want.

**Comment by seabbs** [comment]
Bash script template for automating submission via PR: https://github.com/epiforecasts/covid.german.forecasts/blob/master/submissions/submit.sh

(not plug and play but minimal technical knowledge needed)

**Comment by jbracher** [comment]
I don't have any real stakes here, but we should keep in mind that eventually this will be maintained by other people with very limited time. Keeping things simple and limiting dependencies thus seems like a good idea to me.

That being said I think that the visualization of peoples' forecasts could lead to real improvements in forecast quality. And it does not involve other people running their code in the repo, so I'm in favour of that.

**Comment by seabbs** [comment]
I don't think what we are saying is in conflict @jbracher. My main point is that teams submitting also have people maintaining their forecasts all of whom have very limited time. 

I'm suggesting providing either some tools for people to adapt on their own (i.e link above which runs from a forecasters repository and has nothing to do with hub code) or just a place for them to share those tools with each other. Providing additional small scale functions written by different teams or by any other contributor might be a good way to reduce the burden which at the moment is kept away from the hub members by shifting it onto those submitting.
